### ISSUE TITLE
Segmentation loop

### ISSUE THREAD

**Issue by winglian** [jlamprou/Infini-Attention#1]
> Theory 2: Segment the input during the training loop and feed each segment to the model with gradient accumulation steps equal to Sequence Length / Segment Length.

Looking at the last four lines of https://raw.githubusercontent.com/kimiyoung/transformer-xl/master/pytorch/mem_transformer.py you can see that it's likely splitting the context outside of the model forward.

**Comment by jlamprou** [comment]
@winglian Yep that's what I'm thinking too. My internal implementation does something like that, but i can't make sense of the right training loop. Running CLM training for example, how do we calculate the loss with the input (2048) and labels (whole seq_len), also at which point we perform the backward pass, in each segment or after the last segment? I think for starters we trash the huggingface CausalLM Class which takes care the loss calculation and perform the loss calculation in the training loop. Feel Free to PR the repo if you have any ideas!

**Comment by jlamprou** [comment]
@winglian I updated the repo with a training loop with segmentation and it works!!! Try it!

**Comment by winglian** [comment]
@jlamprou I'll check it out shortly! Have you tried to inference with this yet? One of the things that seems to break down is that if you don't have a sequence length that evenly divides the segment_len, it seems to fail.

**Comment by jlamprou** [comment]
@winglian I haven't checked the inference yet, but padding probably will solve the indivisible sequence lengths or feed the remainder as an extra segment.

**Comment by winglian** [comment]
looks like you might need to pass `M_Z` here when using gradient checkpointing? https://github.com/jlamprou/Infini-Attention/blob/main/modeling_qwen_transformers.py#L1441
